### ISSUE TITLE
chap06_Keras_Model_keras_mnist_DNN_dropout

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_deep_learning#48]
"""
Tensorflow2.0 keras + Mnist(0~9) + Flatten layer + History + Dropout

1차 방법 : 1차원 : (28x28) -> 784
2차 방법 : 2차원 : 28 x 28 -> flatten 적용
"""
"""
-드롭아웃은 정확도는 떨어질수 있지만 발리데이션과 트렌인 사이의 차이를 줄임으로써 오버피팅을 줄여 새로운 데이터가
 왔을 때 일반화에 도움을 준다.
-128개 노드(혹은 파라미터 수)가 많네! -> 드롭아웃 0.5로 크게하자.
-output으로 갈수록 드롭아웃비율 적게함
-output layer 적용 안함
"""

import tensorflow as tf # ver2.x
from tensorflow.keras.datasets.mnist import load_data # ver2.x dataset
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import minmax_scale # x변수 전처리
from tensorflow.keras.utils import to_categorical # 원핫코딩과 비슷한것, y변수 전처리
from tensorflow.keras import Sequential # model 생성
from tensorflow.keras.layers import Dense, Flatten, Dropout # layer 생성
from tensorflow.keras.models import load_model # model save/load
from sklearn.metrics import accuracy_score

### 1. x, y 공급 data 
(x_train, y_train), (x_val, y_val) = load_data()
x_train.shape # image # (60000, 28, 28) # 3없음:색깔없음
y_train.shape # labels # (60000,)

- x변수 전처리 : 정규화
x_train[0] # 0~255숫자가 28x28행렬형태
x_train = x_train/255.  # 255. 으로 나누어줌으로써 실수로 변경
x_val = x_val/255.

'''
- 2d -> 1d
x_train = x_train.reshape(-1, 784)
x_val = x_val.reshape(-1, 784)
'''

- y변수 : one hot encoding
y_train = to_categorical(y_train) # y변수 전처리 : one hot encoding
y_val = to_categorical(y_val)
y_train.shape # (60000, 10)

### 2. keras model 생성
model = Sequential()
model # object info

### 3. model layer
'''
model.add(Dense(node수, input_shape, activation)) # hidden layer1
model.add(Dense(node수, activation)) # hidden layer1 ~ n
'''
input_shape = (28, 28)

- Flatten layer  2d(28,28) -> 1d(784)
model.add(Flatten(input_shape = input_shape)) # 0층

- hidden layer1 = [784,128]
model.add(Dense(128, activation = 'relu')) # 1층
Dropout(rate=0.5) # 노드 중에서 50% 제거
- hidden layer2 = [128,64]
model.add(Dense(64, activation = 'relu')) # 2층
Dropout(rate=0.3) # 노드 중에서 30% 제거
- hidden layer3 = [64,32]
model.add(Dense(32, activation = 'relu')) # 3층
Dropout(rate=0.3) # 노드 중에서 30% 제거
- output layer2 = [32,10]
model.add(Dense(10, activation = 'softmax')) # 4층

### 4. model compile : 학습환경 설정
model.compile(optimizer = 'adam', # 최적화 알고리즘(lr 생략)
              loss = 'categorical_crossentropy', # 손실
              metrics=['accuracy']) # 평가 방법

- layer 확인
model.summary()

### 5. model training : train(112) vs val(38)
model_fit = model.fit(x=x_train, y=y_train,  # 학습용
          epochs=15,
          verbose=1, # 학습과정 횟수
          validation_data=(x_val, y_val) # 평가용
          )
'''
Epoch 15/15
1875/1875 [==============================] - 3s 1ms/step
 - loss: 0.0168 - accuracy: 0.9947 - val_loss: 0.1065 - val_accuracy: 0.9765
'''

### 6. model history
-model.fit의 객체를 받을 수 있는 것이 history 객체
print(model_fit.history.keys())
-dict_keys(['loss', 'accuracy', 'val_loss', 'val_accuracy'])

train_loss = model_fit.history['loss']
train_accuracy = model_fit.history['accuracy']
val_loss = model_fit.history['val_loss']
val_accuracy = model_fit.history['val_accuracy']

- train vs val loss
import matplotlib.pyplot as plt
plt.plot(train_loss, color = 'y', label = 'train loss')
plt.plot(val_loss, color = 'r', label = 'val loss')
plt.legend(loc='best')
plt.xlabel = "epochs"
plt.show()
두 격차가 클수록 오버피팅이 우려되므로 2epoch 정도까지만 학습하는 것이 좋겠다.

- train vs val accuracy
import matplotlib.pyplot as plt
plt.plot(train_accuracy, color = 'y', label = 'train acc')
plt.plot(val_accuracy, color = 'r', label = 'val acc')
plt.legend(loc='best')
plt.xlabel = "epochs"
plt.show()
-두 격차가 클수록 오버피팅이 우려되므로 2epoch 정도까지만 학습하는 것이 좋겠다.
